gpio/src/gpio.rs

```diff
@@ -0,0 +1,943 @@
+    GpioIrqTypeNoChange(u16),
+    #[error("Gpio irq not enabled yet")]
+    GpioIrqNotEnabled,
+    #[error(
+        "Current Gpio irq type is valid, must configure to VIRTIO_GPIO_IRQ_TYPE_NONE first {0}"
+    )]
+    GpioOldIrqTypeValid(u16),
+    #[error("Gpio line-request not configured")]
+    GpioLineRequestNotConfigured,
+    #[cfg(test)]
+    #[error("Gpio test Operation failed {0}")]
+    GpioOperationFailed(&'static str),
+}
+
+/// Virtio specification definitions
+/// Virtio GPIO request types
+pub(crate) const VIRTIO_GPIO_MSG_GET_LINE_NAMES: u16 = 0x0001;
+pub(crate) const VIRTIO_GPIO_MSG_GET_DIRECTION: u16 = 0x0002;
+pub(crate) const VIRTIO_GPIO_MSG_SET_DIRECTION: u16 = 0x0003;
+pub(crate) const VIRTIO_GPIO_MSG_GET_VALUE: u16 = 0x0004;
+pub(crate) const VIRTIO_GPIO_MSG_SET_VALUE: u16 = 0x0005;
+pub(crate) const VIRTIO_GPIO_MSG_IRQ_TYPE: u16 = 0x0006;
+
+/// Direction types
+pub(crate) const VIRTIO_GPIO_DIRECTION_NONE: u8 = 0x00;
+pub(crate) const VIRTIO_GPIO_DIRECTION_OUT: u8 = 0x01;
+pub(crate) const VIRTIO_GPIO_DIRECTION_IN: u8 = 0x02;
+
+/// Virtio GPIO IRQ types
+pub(crate) const VIRTIO_GPIO_IRQ_TYPE_NONE: u16 = 0x00;
+pub(crate) const VIRTIO_GPIO_IRQ_TYPE_EDGE_RISING: u16 = 0x01;
+pub(crate) const VIRTIO_GPIO_IRQ_TYPE_EDGE_FALLING: u16 = 0x02;
+pub(crate) const VIRTIO_GPIO_IRQ_TYPE_EDGE_BOTH: u16 =
+    VIRTIO_GPIO_IRQ_TYPE_EDGE_RISING | VIRTIO_GPIO_IRQ_TYPE_EDGE_FALLING;
+pub(crate) const VIRTIO_GPIO_IRQ_TYPE_LEVEL_HIGH: u16 = 0x04;
+pub(crate) const VIRTIO_GPIO_IRQ_TYPE_LEVEL_LOW: u16 = 0x08;
+const VIRTIO_GPIO_IRQ_TYPE_ALL: u16 = VIRTIO_GPIO_IRQ_TYPE_EDGE_BOTH
+    | VIRTIO_GPIO_IRQ_TYPE_LEVEL_HIGH
+    | VIRTIO_GPIO_IRQ_TYPE_LEVEL_LOW;
+
+/// Virtio GPIO Configuration
+#[derive(Clone, Debug, PartialEq)]
+#[repr(C)]
+pub(crate) struct VirtioGpioConfig {
+    pub(crate) ngpio: Le16,
+    pub(crate) padding: Le16,
+    pub(crate) gpio_names_size: Le32,
+}
+
+/// Trait that represents an GPIO Device.
+///
+/// This trait is introduced for development purposes only, and should not
+/// be used outside of this crate. The purpose of this trait is to provide a
+/// mock implementation for the GPIO driver so that we can test the GPIO
+/// functionality without the need of a physical device.
+pub(crate) trait GpioDevice: Send + Sync + 'static {
+    fn open(device: u32) -> Result<Self>
+    where
+        Self: Sized;
+
+    fn get_num_lines(&self) -> Result<u16>;
+    fn get_line_name(&self, gpio: u16) -> Result<String>;
+    fn get_direction(&self, gpio: u16) -> Result<u8>;
+    fn set_direction(&self, gpio: u16, dir: u8, value: u32) -> Result<()>;
+    fn get_value(&self, gpio: u16) -> Result<u8>;
+    fn set_value(&self, gpio: u16, value: u32) -> Result<()>;
+
+    fn set_irq_type(&self, gpio: u16, value: u16) -> Result<()>;
+    fn wait_for_interrupt(&self, gpio: u16) -> Result<()>;
+}
+
+pub(crate) struct PhysDeviceState {
+    // See wait_for_interrupt() for explanation of Arc.
```

Comment:
I think this should be "PhysLineState".  Otherwise it is very confusing when considering that "PhysDevice" below holds a vector of those.  It is much easier to think about PhysDevice as an aggregation of a GPIO "Chip" with "nlines" (rather than "ngpio"), where the state of each line is a vector of "PhysLineState".